app/controllers/documents_v2_controller.rb

```diff
@@ -0,0 +1,119 @@
+class DocumentsV2Controller < ApplicationController
+
+  before_filter :require_valid_access_key, :only => [:open]
+  before_filter :require_valid_read_write_access_key, :only => [:save, :patch]
+  before_filter :require_valid_shared_recordid, :only => [:copy_shared]
+
+  # don't check for csrf token
+  skip_before_filter :verify_authenticity_token, :only => [:save, :patch, :copy_shared]
+
+  # v2 of the api is all anonymous access
+  skip_authorization_check
+
+  include DocumentsV2Helper
+
+  def open
+    readOnly = params[:readAccessKey].present?
+    response.headers['Document-Id'] = "#{@document.id}"
+    response.headers['Allow'] = "GET, HEAD, OPTIONS#{readOnly ? '' : ', PUT, PATCH'}"
+    response.headers['X-Document-Store-Read-Only'] = readOnly ? 'true' : 'false'
+    render json: @document.content
+  end
+
+  def save
+    @document.form_content = request.raw_post
+    if @document.save
+      render json: {status: "Saved", valid: true, id: @document.id }, status: 200
+    else
+      render json: {status: "Error", errors: @document.errors.full_messages, valid: false, message: 'error.writeFailed' }, status: 400
+    end
+  end
+
+  def patch
+    begin
+      patchset = JSON.parse(request.raw_post)
+      raise "Empty patchset" unless patchset && patchset.is_a?(Array)
+    rescue => e
+      render json: {status: "Error", errors: ["Invalid patch JSON (parsing)", e.to_s, patchset], valid: false, message: 'error.writeFailed' }, status: 400
+      return
+    end
+
+    begin
+      patchedContent = JSON::Patch.new(@document.content, patchset).call
+      if @document.update_columns({updated_at: Time.current}) && @document.contents.update_columns({content: patchedContent, updated_at: Time.current})
+        render json: {status: "Patched", valid: true, id: @document.id}, status: 200
+      else
+        render json: {status: "Error", errors: @document.errors.full_messages + @document.contents.errors.full_messages, valid: false, message: 'error.writeFailed' }, status: 400
+      end
+    rescue => e
+      render json: {status: "Error", errors: ["Invalid patch JSON (executing)", e.to_s], valid: false, message: 'error.writeFailed' }, status: 400
+    end
+  end
+
+  def copy_shared
```

Comment:
Same thing here as above. It seems more clear to just inline the code for `require_valid_shared_recordid`.